In [1]:
# Dependencies
import pandas as pd # type: ignore
from pathlib import Path
import pyreadstat  # type: ignore

## Reading in Datasets & Basic Data Exploration:

### Prescription Medications: 

In [2]:
Prescription_Meds = pd.read_sas("Datasets\Prescription_meds.xpt")
Prescription_Meds.head()

SEQN  RXDUSE   RXDDRUG RXDDRGID  RXQSEEN  RXDDAYS RXDRSC1 RXDRSC2  \
0  109263.0     2.0       b''      b''      NaN      NaN     b''     b''   
1  109264.0     2.0       b''      b''      NaN      NaN     b''     b''   
2  109265.0     2.0       b''      b''      NaN      NaN     b''     b''   
3  109266.0     2.0       b''      b''      NaN      NaN     b''     b''   
4  109267.0     1.0  b'99999'      b''      NaN      NaN     b''     b''   

  RXDRSC3 RXDRSD1 RXDRSD2 RXDRSD3  RXDCOUNT  
0     b''     b''     b''     b''       NaN  
1     b''     b''     b''     b''       NaN  
2     b''     b''     b''     b''       NaN  
3     b''     b''     b''     b''       NaN  
4     b''     b''     b''     b''       1.0

In [3]:
#df = pyreadstat.read_xport('Datasets\Prescription_meds.xpt', metadataonly=True)
Prescription_Meds.info()
print(f"Rows, Columns: {Prescription_Meds.shape}")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32962 entries, 0 to 32961
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   SEQN      32962 non-null  float64
 1   RXDUSE    32962 non-null  float64
 2   RXDDRUG   32962 non-null  object 
 3   RXDDRGID  32962 non-null  object 
 4   RXQSEEN   24037 non-null  float64
 5   RXDDAYS   24031 non-null  float64
 6   RXDRSC1   32962 non-null  object 
 7   RXDRSC2   32962 non-null  object 
 8   RXDRSC3   32962 non-null  object 
 9   RXDRSD1   32962 non-null  object 
 10  RXDRSD2   32962 non-null  object 
 11  RXDRSD3   32962 non-null  object 
 12  RXDCOUNT  24246 non-null  float64
dtypes: float64(5), object(8)
memory usage: 3.3+ MB
Rows, Columns: (32962, 13)


### Physical Activity:

In [4]:
Physical_Activity = pd.read_sas("Datasets\Phys_Activity.xpt")
Physical_Activity.head()

SEQN  PAQ605  PAQ610  PAD615  PAQ620  PAQ625  PAD630  PAQ635  PAQ640  \
0  109266.0     2.0     NaN     NaN     2.0     NaN     NaN     2.0     NaN   
1  109267.0     2.0     NaN     NaN     2.0     NaN     NaN     2.0     NaN   
2  109268.0     1.0     5.0   540.0     1.0     5.0   300.0     2.0     NaN   
3  109271.0     2.0     NaN     NaN     1.0     2.0   120.0     2.0     NaN   
4  109273.0     1.0     3.0   240.0     2.0     NaN     NaN     2.0     NaN   

   PAD645  PAQ650  PAQ655  PAD660  PAQ665  PAQ670  PAD675  PAD680  
0     NaN     1.0     5.0    60.0     1.0     4.0    30.0   480.0  
1     NaN     1.0     6.0    90.0     2.0     NaN     NaN   540.0  
2     NaN     2.0     NaN     NaN     2.0     NaN     NaN   540.0  
3     NaN     2.0     NaN     NaN     2.0     NaN     NaN    60.0  
4     NaN     2.0     NaN     NaN     1.0     4.0   120.0   180.0

In [5]:
Physical_Activity.info()
print(f"Rows, Columns: {Physical_Activity.shape}")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9693 entries, 0 to 9692
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   SEQN    9693 non-null   float64
 1   PAQ605  9693 non-null   float64
 2   PAQ610  2418 non-null   float64
 3   PAD615  2402 non-null   float64
 4   PAQ620  9693 non-null   float64
 5   PAQ625  4187 non-null   float64
 6   PAD630  4164 non-null   float64
 7   PAQ635  9693 non-null   float64
 8   PAQ640  2253 non-null   float64
 9   PAD645  2241 non-null   float64
 10  PAQ650  9693 non-null   float64
 11  PAQ655  2422 non-null   float64
 12  PAD660  2417 non-null   float64
 13  PAQ665  9693 non-null   float64
 14  PAQ670  3904 non-null   float64
 15  PAD675  3893 non-null   float64
 16  PAD680  9676 non-null   float64
dtypes: float64(17)
memory usage: 1.3 MB
Rows, Columns: (9693, 17)


### Demographics Data:

In [6]:
Demographics = pd.read_sas("Datasets\Demograpics.xpt")
Demographics.head()

SEQN  SDDSRVYR  RIDSTATR  RIAGENDR  RIDAGEYR  RIDAGEMN  RIDRETH1  \
0  109263.0      66.0       2.0       1.0       2.0       NaN       5.0   
1  109264.0      66.0       2.0       2.0      13.0       NaN       1.0   
2  109265.0      66.0       2.0       1.0       2.0       NaN       3.0   
3  109266.0      66.0       2.0       2.0      29.0       NaN       5.0   
4  109267.0      66.0       1.0       2.0      21.0       NaN       2.0   

   RIDRETH3  RIDEXMON  DMDBORN4  ...  FIAINTRP  MIALANG  MIAPROXY  MIAINTRP  \
0       6.0       2.0       1.0  ...       2.0      NaN       NaN       NaN   
1       1.0       2.0       1.0  ...       2.0      1.0       2.0       2.0   
2       3.0       2.0       1.0  ...       2.0      NaN       NaN       NaN   
3       6.0       2.0       2.0  ...       2.0      1.0       2.0       2.0   
4       2.0       NaN       2.0  ...       2.0      NaN       NaN       NaN   

   AIALANGA      WTINTPRP      WTMECPRP  SDMVPSU  SDMVSTRA  INDFMPIR  
0       NaN   7891.762435  8.951816e+03      3.0     156.0      4.66  
1       1.0  11689.747264  1.227116e+04      1.0     155.0      0.83  
2       NaN  16273.825939  1.665876e+04      1.0     157.0      3.06  
3       1.0   7825.646112  8.154968e+03      2.0     168.0      5.00  
4       NaN  26379.991724  5.397605e-79      1.0     156.0      5.00  

[5 rows x 29 columns]

In [7]:
Demographics.info()
print(f"Rows, Columns: {Demographics.shape}")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15560 entries, 0 to 15559
Data columns (total 29 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   SEQN      15560 non-null  float64
 1   SDDSRVYR  15560 non-null  float64
 2   RIDSTATR  15560 non-null  float64
 3   RIAGENDR  15560 non-null  float64
 4   RIDAGEYR  15560 non-null  float64
 5   RIDAGEMN  987 non-null    float64
 6   RIDRETH1  15560 non-null  float64
 7   RIDRETH3  15560 non-null  float64
 8   RIDEXMON  14300 non-null  float64
 9   DMDBORN4  15560 non-null  float64
 10  DMDYRUSZ  3028 non-null   float64
 11  DMDEDUC2  9232 non-null   float64
 12  DMDMARTZ  9232 non-null   float64
 13  RIDEXPRG  1874 non-null   float64
 14  SIALANG   15560 non-null  float64
 15  SIAPROXY  15560 non-null  float64
 16  SIAINTRP  15560 non-null  float64
 17  FIALANG   14481 non-null  float64
 18  FIAPROXY  14481 non-null  float64
 19  FIAINTRP  14481 non-null  float64
 20  MIALANG   11000 non-null  fl

### Sleep Disorders: 

In [8]:
Sleep_Disorders.info()
print(f"Rows, Columns: {Sleep_Disorders.shape}")

NameError: name 'Sleep_Disorders' is not defined

## Initial Observations and concerns:

The datasets appear to vary widely in the number of rows/observations the contain, ranging from 32962 rows (Prescription_Meds) down to 9693 rows (Physical_Activity). Things are further complicated by the fact that we plan to merge the datasets based on the SEQN column values (respondant sequence numbers) in order to match up specific respondants across datasets. Thus, if we want to analyze data from a column in Physical_Activity alongside data from the other larger datasets, we will either be left with a bunch of missing/empty values or lose quite a bit of our sample size. A partial remedy to this could be creating several merged datasets that are unique to a specific analysis we want to preform. Similarly, we could just pull out specific columns of interest and create a seperate dataframe for an analysis. Lastly, there are also a number of variables that are coded in a convoluted way, as is usually the case with questionaire data. In light of these considerations it is clear that we will have quite a bit of data wranging and cleaning to do before we can begin our analyses.

I am going to merge all of the datasets together below, making sure to use a left join to retain the max number of rows (32962) even though there will be many empty rows



## Initial Merge of Prescription_Meds, Demographics, Physical_Activity, Sleep_Disorders:

### Left Join 

In [ ]:
# Beginning first merge: Prescription_Meds, Demographics,
First_Merge = pd.merge(Prescription_Meds, Demographics, on=["SEQN"], how = 'left')
print(f"First merge: {First_Merge.shape}")  

# Second merge to add Physical_Activity:
Second_Merge = pd.merge(First_Merge, Physical_Activity, on=["SEQN"], how = 'left')
print(f"Second merge: {Second_Merge.shape}")  

# Final merge to add Sleep_Disorders:
Merged_Data = pd.merge(Second_Merge, Sleep_Disorders, on=["SEQN"], how = 'left')
print(f"Final merge: {Merged_Data.shape}")  

## Exploring the Merged Data

In [ ]:
Merged_Data.info()

#### As stated previously, we can always modify how we manipulate specific data/columns to preseve data later on. 

## Exporting the Merged Dataframe into a CSV File

In [ ]:
# Push the remade DataFrame to a new CSV file
Merged_Data.to_csv("Datasets/Merged_Dataset",
                  encoding="utf-8", index=False, header=True)